### ISSUE TITLE
Unable to swipe between slides

### ISSUE THREAD

**Issue by romancampula** [AppIntro/AppIntro#347]
**AppIntro Version**:  4.1.0

**Device/Android Version**: every device I have tried (e.g. Samsung Galaxy S5mini, LG H440N, any in  emulator)

**Issue details / Repro steps / Use case background**:
In intro with permissions it is necessary to use e.g.
`android.Manifest.permission.ACCESS_FINE_LOCATION`
instead of
`Manifest.permission.ACCESS_FINE_LOCATION`

But adding `android` causes that I can't swipe between slides. When `android` is removed, i can swipe but than it can't resolve symbol ACCESS_FINE_LOCATION.

This bad behavior is also in example project AppIntroExample on Google Play.

**Comment by avluis** [comment]
What happened to;

**Your Code:**

**Stack trace / LogCat:**

**Comment by romancampula** [comment]
If you really need it, here it is:

**Your Code:**
`package cz.cdv.modi;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;

import com.github.paolorotolo.appintro.AppIntro;

public class IntroActivity extends AppIntro {
    ////////////////////////////////////////////////////////////////////////////////////////////////
    //                              FUNKCE ZIVOTNIHO CYKLU APLIKACE                               //
    ////////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // actionbar
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(false);
        }

        // ziskani nastaveni aplikace
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);

        // ulozeni hodnoty firstStart do nastaveni aplikace
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putBoolean("firstStart", false);
        editor.commit();

        // pridani slidu
        Fragment slide1Fragment = new Slide1Fragment();
        Fragment slide2Fragment = new Slide2Fragment();
        Fragment slide3Fragment = new Slide3Fragment();
        Fragment slide4Fragment = new Slide4Fragment();
        Fragment slide5Fragment = new Slide5Fragment();

        addSlide(slide1Fragment);
        addSlide(slide2Fragment);
        addSlide(slide3Fragment);
        addSlide(slide4Fragment);
        addSlide(slide5Fragment);

        // nastaveni
        showSkipButton(false);
        setProgressButtonEnabled(true);
        setDoneText(getResources().getString(R.string.done));
        setBarColor(ContextCompat.getColor(this, R.color.colorPrimary));
        setSeparatorColor(ContextCompat.getColor(this, R.color.colorPrimaryDark));

        // zajisteni opravneni
        askForPermissions(new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION,
                                       android.Manifest.permission.ACCESS_WIFI_STATE,
                                       android.Manifest.permission.CHANGE_WIFI_STATE,
                                       android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, 2);
    }

    /**
     * Akce po stisknuti tlacitka "preskocit"
     */
    @Override
    public void onSkipPressed(Fragment currentFragment) {
        super.onSkipPressed(currentFragment);

        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
    }

    /**
     * Akce po stisknuti tlacitka "hotovo"
     */
    @Override
    public void onDonePressed(Fragment currentFragment) {
        super.onDonePressed(currentFragment);

        Intent intent = new Intent(this, MainActivity.class);
        startActivity(intent);
    }

    /**
     * Akce pri zmene slidu
     */
    @Override
    public void onSlideChanged(@Nullable Fragment oldFragment, @Nullable Fragment newFragment) {
        super.onSlideChanged(oldFragment, newFragment);

        // ziskani nastaveni aplikace
        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);

        if (oldFragment != null) {
            EditText editText_userId = (EditText) findViewById(R.id.editText_userId);
            if (oldFragment.getView() == findViewById(R.id.intro_slide3_fragment)) {
                // ulozeni userId do nastaveni aplikace
                if (editText_userId.getText().toString().length() > 0) {
                    SharedPreferences.Editor editor = sharedPreferences.edit();
                    editor.putString("userId", editText_userId.getText().toString());
                    editor.commit();
                }
                else {
                    SharedPreferences.Editor editor = sharedPreferences.edit();
                    editor.putString("userId", "0");
                    editor.commit();
                }
            }
        }
    }

    /**
     * Fragment slide 1
     */
    public static class Slide1Fragment extends Fragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.intro_slide1_fragment, container, false);
        }
    }

    /**
     * Fragment slide 2
     */
    public static class Slide2Fragment extends Fragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.intro_slide2_fragment, container, false);
        }
    }

    /**
     * Fragment slide 3
     */
    public static class Slide3Fragment extends Fragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.intro_slide3_fragment, container, false);
        }
    }

    /**
     * Fragment slide 4
     */
    public static class Slide4Fragment extends Fragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.intro_slide4_fragment, container, false);
        }
    }

    /**
     * Fragment slide 5
     */
    public static class Slide5Fragment extends Fragment {
        @Override
        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            return inflater.inflate(R.layout.intro_slide5_fragment, container, false);
        }
    }
}`

**Stack trace / LogCat:**
11-04 16:09:02.389 cz.cdv.modi V/ViewRootImpl: Contents drawing finished : cz.cdv.modi/cz.cdv.modi.MainActivity
11-04 16:09:02.437 cz.cdv.modi V/ViewRootImpl: Contents drawing finished : cz.cdv.modi/cz.cdv.modi.IntroActivity
11-04 16:09:02.481 cz.cdv.modi I/Timeline: Timeline: Activity_idle id: android.os.BinderProxy@726f634 time:6259945
11-04 16:09:02.481 cz.cdv.modi I/Timeline: Timeline: Activity_idle id: android.os.BinderProxy@150950d time:6259945
11-04 16:10:20.787 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_DOWN
11-04 16:10:21.780 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_UP
11-04 16:10:33.829 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_DOWN
11-04 16:10:34.205 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_UP
11-04 16:10:40.953 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_DOWN
11-04 16:10:41.005 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_UP
11-04 16:10:41.014 cz.cdv.modi I/AudioManagerEx: AudioManagerEx created
11-04 16:10:44.854 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_DOWN
11-04 16:10:44.999 cz.cdv.modi I/ViewRootImpl: ViewRoot's Touch Event : ACTION_UP

**Comment by romancampula** [comment]
Some notes from testing:

- It is possible to swipe on Samgung Galaxy S5 mini (SM-G800F, Android 5.1.1)
- It is possible to swipe on LG Spirit G4 LTE (LG-H440n, Android 5.0.1)
- It is not possible to swipe on LG Spirit G4 LTE (LG-H440n, Android 6.0)
- It is not possible to swipe on emulator (Android 6.0)

**Comment by avluis** [comment]
Let's try and work from top to bottom -- I'm seeing a few things that need to be fixed before the issue itself is apparent.

- Check out this [sample fragment](https://github.com/paolorotolo/AppIntro/blob/master/example/src/main/java/com/github/paolorotolo/appintroexample/SampleSlide.java) -- use that to init your fragments (remove your init code as well).
- The fragment above expects to find several ids in your layout: make use of the [provided](https://github.com/paolorotolo/AppIntro/tree/master/example/src/main/res/layout) AppIntro and AppIntro2 layouts to build your own.

Taking a wild stab from the logcat above -- the viewpager does not see the id to attach to -- not sure if this is entirely responsible for your testing results, but do note that swyping is supposed to be disabled when permissions are being requested.

**Comment by pcegarra** [comment]
When I check for permissions with this line :

`        askForPermissions(new String[]{Manifest.permission.CAMERA}, 2);
`
the intro dont allow swipe gestures :(

**Comment by Ascendance** [comment]
This is an issue can confirm. Disabling swipe defeats the purpose of all the built-in transition animations.

**Comment by shmuelr** [comment]
I can confirm this issue. Adding a permission to the onboarding in Android 6.0 and higher disables the swiping

**Comment by shmuelr** [comment]
This is the source of the bug:
https://github.com/paolorotolo/AppIntro/blob/master/library/src/main/java/com/github/paolorotolo/appintro/AppIntroBase.java#L804

    public void askForPermissions(String[] permissions, int slidesNumber) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (slidesNumber == 0) {
                Toast.makeText(getBaseContext(), "Invalid Slide Number", Toast.LENGTH_SHORT).show();
            } else {
                PermissionObject permission = new PermissionObject(permissions, slidesNumber);
                permissionsArray.add(permission);
                setSwipeLock(true);
            }
        }
    }

On M+ devices swiping gets disabled if we need to ask the user for permission

**Comment by Andrew-Quebe** [comment]
@pcegarra @Ascendance @shmuelr the reason swiping is disabled is because of the way Android's viewpager listener works. The request MUST be made from a button press vs when the sliding of the viewpager occurs. 

When @paolorotolo and I were implementing this, we actually ran into the issue of the request being made tons of times based on how far we made the slide move without using the button. 

If someone knows of a way to request the permissions ONCE when the user begins sliding, then we can allow swiping if a user requests permissions. Until then, that is how Android works.

**Comment by Andrew-Quebe** [comment]
@shmuelr to add to the above, it's not a bug...it's simply a workaround we had to do to make sure the request was not made a bunch of times. If you don't believe me, compile the source yourself and remove that line...you'll notice what I'm talking about.

**Comment by fruel** [comment]
@Andrew-Quebe I stumbled upon this swiping issue/workaround just now and managed to write my own workaround using ISlidePolicy. 

I'm not sure if I missed something but in my initial 5min testing I did not encounter that the permission request is "being made tons of times". The permission dialog is shown when I try to leave the fragment via the "Next"  button or by starting to scroll. When I allow or deny the permission request the sliding to the next page continues. The only downside I encountered so far is that the request will also show up when scrolling backwards to the previous tile.

Tested on an Oneplus 3 with Android 6.0.1

My implementation is a small wrapper around AppIntro2Fragment:
```java
public class PermissionAppIntroFragment extends AppIntroBaseFragment implements ISlidePolicy {

    public static PermissionAppIntroFragment newInstance(CharSequence title, CharSequence description, @DrawableRes int imageDrawable, @ColorInt int bgColor) {
        // [copy from AppIntro2Fragment]
    }

    @Override
    protected int getLayoutId() {
        return com.github.paolorotolo.appintro.R.layout.fragment_intro2;
    }

    @Override
    public boolean isPolicyRespected() {
        return ContextCompat.checkSelfPermission(getContext(), Manifest.permission.READ_CONTACTS) == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void onUserIllegallyRequestedNextPage() {
        Log.d("AppIntro", "Permission request");
        ActivityCompat.requestPermissions(getActivity(), new String[]{ Manifest.permission.READ_CONTACTS }, 1);
    }
}
```

**Comment by avluis** [comment]
@fruel Can you test the following:
Put several permissions on the same slide like so:
`askForPermissions(new String[]{Manifest.permission.CAMERA, Manifest.permission.READ_CONTACTS}, 2);`

Let us know if your wrapper has no issue with the above~

**Comment by fruel** [comment]
@avluis 
Working fine. Behaves exactly the same. 

Scrolling backwards behaves a bit weird (but still the same as with one permission): It will reverse direction and scroll forward when the permission dialog closes.

```java
    @Override
    public boolean isPolicyRespected() {
        return ContextCompat.checkSelfPermission(getContext(), Manifest.permission.READ_CONTACTS) == PackageManager.PERMISSION_GRANTED
                && ContextCompat.checkSelfPermission(getContext(), Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void onUserIllegallyRequestedNextPage() {
        Log.d("AppIntro", "Permission request");
        ActivityCompat.requestPermissions(getActivity(), new String[]{Manifest.permission.CAMERA,  Manifest.permission.READ_CONTACTS }, 1);
    }
```

**Comment by fruel** [comment]
Here a snipped of my logcat. After the permission request is started the Activity is paused. When the dialog closes it is resumed and scrolling continues.

```
11-29 00:59:15.256 22951-22951/com.test.package D/AppIntro: Permission request    <--- Log.d() from above
11-29 00:59:15.281 22951-22951/com.test.package D/AppTracker: App Event: stop
11-29 00:59:15.281 22951-22980/com.test.package V/FA: Recording user engagement, ms: 2203
11-29 00:59:15.282 22951-22980/com.test.package V/FA: Activity paused, time: 1060292126
11-29 00:59:17.575 22951-22951/com.test.package D/AppTracker: App Event: start
11-29 00:59:17.576 22951-22980/com.test.package V/FA: Activity resumed, time: 1060294421
```

**Comment by Andrew-Quebe** [comment]
@fruel is there any hint of scrolling when the dialog pops up? The other issue with having swiping enabled is that you may slide a small amount, and (quite literally) the perm dialog will show under your fingers...this freezes the UI underneath. Does this happen with your workaround?

In theory, the process of the intro needs to be as follows:
First slide --> Second slide with permission INFO --> Request second slide permission WITHOUT the pager sliding --> Proceed to slide 3 AFTER permission dialog has closed (permission acceptance or denial is irrelevant). --> etc. etc.

**Comment by Andrew-Quebe** [comment]
Basically, we need to be able to have a smooth UI/UX, while still requesting permissions.

**Comment by Andrew-Quebe** [comment]
Label all the things :P

**Comment by fruel** [comment]
![](https://thumbs.gfycat.com/GlisteningIllElephantbeetle-size_restricted.gif)
(Video: https://gfycat.com/GlisteningIllElephantbeetle)

**Comment by fruel** [comment]
And as reference - this is my complete App Intro Activity:

```java
public class TutorialActivity extends AppIntro2 {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        addSlide(AppIntro2Fragment.newInstance(getString(R.string.tutorial_page_1_title), getString(R.string.tutorial_page_1_description), R.drawable.placeholder, ContextCompat.getColor(this, R.color.tutorialPage1)));
        addSlide(AppIntro2Fragment.newInstance(getString(R.string.tutorial_page_2_title), getString(R.string.tutorial_page_2_description), R.drawable.placeholder, ContextCompat.getColor(this, R.color.tutorialPage2)));
        addSlide(PermissionAppIntroFragment.newInstance(getString(R.string.tutorial_page_3_title), getString(R.string.tutorial_page_3_description), R.drawable.placeholder, ContextCompat.getColor(this, R.color.tutorialPage3)));
        addSlide(AppIntro2Fragment.newInstance(getString(R.string.tutorial_page_4_title), getString(R.string.tutorial_page_4_description), R.drawable.placeholder, ContextCompat.getColor(this, R.color.tutorialPage4)));

        showSkipButton(false);

        setColorTransitionsEnabled(true);

    }

    @Override
    public void onDonePressed(Fragment currentFragment) {
        super.onDonePressed(currentFragment);

   
    }
}

```

**Comment by voghDev** [comment]
Don't know if it's too late but, having a look at the askForPermissions method:

    public void askForPermissions(String[] permissions, int slidesNumber) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (slidesNumber == 0) {
                Toast.makeText(getBaseContext(), "Invalid Slide Number", Toast.LENGTH_SHORT).show();
            } else {
                PermissionObject permission = new PermissionObject(permissions, slidesNumber);
                permissionsArray.add(permission);
                setSwipeLock(true);
            }
        }
    }

You can notice the `setSwipeLock(true);` call. Setting the property back to false after the call seems to work well. Code:

    class OnboardingActivity : AppIntro2() {
        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)        

            // ...

        askForPermissions(arrayOf(Manifest.permission.CAMERA), 2)
        setSwipeLock(false)
        }
    }

**Comment by paolorotolo** [comment]
Hey @voghDev! It's not too late. I didn't had much time to update AppIntro recently.

Let's see if we can fix this at library level. I don't think the user should be the one responsible to make that setSwipeLock call.

P.S. Also kudos for using Kotlin! ;)

**Comment by carstenhag** [comment]
Any chances this is coming any soon? Just noticed this too and wondered that it has been open for so long 👍

**Comment by Anutha02** [comment]
@paolorotolo Any updates on this?

**Comment by Anutha02** [comment]
@cortinico Add this to the Triage as high priority. I am working on this and will submit a PR soon.

**Comment by Anutha02** [comment]
Almost 90% done. Will send a PR soon™.😊

**Comment by Anutha02** [comment]
99% done.

**Comment by Anutha02** [comment]
Also, @paolorotolo This wasn't as simple as only disabling the setSwipelock call....😅😅

**Comment by cortinico** [comment]
So what's the status @Anutha02 here?

**Comment by Anutha02** [comment]
Solved by #654

**Comment by carstenhag** [comment]
#654 may fix this, but it's not merged yet. Imo should be open still :)

**Comment by cortinico** [comment]
@carstenhag Yeah you're right. I confused #654 with #647

**Comment by AnuthaDev** [comment]
@cortinico This issue has been fixed by #678

**Comment by carstenhag** [comment]
Hey, @AnuthaDev, so how can I use this? I am using 6.0.0-alpha2, but I still can not swipe when a permission will be asked.

**Comment by cortinico** [comment]
@carstenhag So this issue has been fixed, merged and is now on master.
I've also just published a tag for AppIntro `6.0.0-alpha3` that contains the patch #678 

If you find a bug with that implementation, please open another issue. I'm closing this one 👍
